crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3245,16 +3270,37 @@ impl<'db> TypeInferenceBuilder<'db> {
                         Type::Unknown
                     })
             }
+            // Ex) Given `"value"[1:3]`, return `"al"`
+            (Type::StringLiteral(literal_ty), Type::SliceLiteral(slice_ty)) => {
+                let literal_value = literal_ty.value(self.db);
+                let start = slice_ty.start(self.db);
+                let stop = slice_ty.stop(self.db);
+                let step = slice_ty.step(self.db);
+                let chars: Vec<_> = literal_value.chars().collect();
```

Comment:
This additional allocation is unfortunate, but `.chars()` is not an `ExactSizeIterator` (Unicode…), and I currently don't see how we can implement the most general form of slicing (e.g. something like `string_literal[3:-5]`) on `DoubleEndedIterator` alone. I could avoid the allocation by computing the length upfront (`O(n)`), but I'm not sure if it's worth investing more time into this.

Reply:
I think this is fine, considering that slices into strings should be rare. We can optimize if this shows up in profiles.